proposals/0000-integrations.md

```diff
@@ -0,0 +1,171 @@
+- Nuxt: [1 step](https://partytown.builder.io/nuxt)
+
+Tailwind suffers from a similar difficult setup story in Astro.
+
+# Detailed design
+
+**Background:** This API was reached through weeks of experimentation of different designs (see alternatives below). To test the work, I designed and build the following integrations, which are useful for illustrating this RFC: 
+
+- **Renderers:** [`lit`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/lit/index.js), [`svelte`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/svelte/index.js), [`react`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/react/index.js), [`preact`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/preact/index.js), [`vue`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/vue/index.js), [`solid`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/solid/index.js)
+- **Libraries:** [`tailwind`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/tailwind/index.js), [`partytown`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/partytown/index.js), [`turbolinks`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/turbolinks/index.js)
+- **Features:** [`sitemap`](https://github.com/withastro/astro/blob/wip-integrations-4/packages/integrations/sitemap/index.js)
+
+## Integration Usage API
+
+```js
+// astro.config.js
+export default ({
+	integrations: [
+		import('@astrojs/vue'),
+        // or, with options
+		[import('@astrojs/vue'), {/* options */}],
+        // or, as a static ESM import at the top of the file
+		vuePlugin({/* options */})
+	],
+});
+```
+
+## Integration Author API
+
+```ts
+export interface AstroIntegration {
+	name: string;
+	hooks: {
+		/** SETUP */
+		/** Called on astro startup, lets you modify config */
+		'astro:config:setup': (options: ConfigSetupOptions) => void | Promise<void>;
+		/** Called after config is finalized, lets you store config object for later */
+		'astro:config:done': (options: { config: Readonly<AstroConfig> }) => void | Promise<void>;
+
+		/** DEV */
+		/** Called on server setup, lets you modify the server */
+		'astro:server:setup': (options: { server: vite.ViteDevServer }) => void | Promise<void>;
+		/** Called on server startup, lets you read the dev server address/URL */
+		'astro:server:start': (options: { address: AddressInfo }) => void | Promise<void>;
+		/** Called on server exit */
+		'astro:server:done': () => void | Promise<void>;
+
+		/** BUILD */
+		/** Called on build start, lets you modify the server */
+		'astro:build:start': () => void | Promise<void>;
+		/** Called on build done, lets you read metadata about the build */
+		'astro:build:done': (options: { pages: string[]; dir: URL }) => void | Promise<void>;
```

Comment:
A common issue I have faced when writing Vite plugins is that hooks are run multiple times. Will these hooks too get called multiple times?

Reply:
Yes, that's 100% a commitment here. That's a great example of why having our own integration system is important, since we can't make guarantees about how Vite plugins are run since we aren't the ones running them.